lld/test/ELF/aarch64-reloc-pauth.s

```diff
@@ -0,0 +1,114 @@
+## but contains only the relative relocations.
+## Any relative relocations with odd offset stay in SHT_RELA.
+
+# RELR:      Section ({{.+}}) .rela.dyn {
+# RELR-NEXT:   0x30430 R_AARCH64_AUTH_RELATIVE - 0x123456789A
+# RELR-NEXT:   0x30449 R_AARCH64_AUTH_RELATIVE - 0x4
+# RELR-NEXT:   0x30438 R_AARCH64_AUTH_ABS64 zed2 0x0
+# RELR-NEXT:   0x30440 R_AARCH64_AUTH_ABS64 bar2 0x0
+# RELR-NEXT: }
+# RELR-NEXT: Section ({{.+}}) .relr.auth.dyn {
+# RELR-NEXT:   0x30410 R_AARCH64_RELATIVE -
+# RELR-NEXT:   0x30418 R_AARCH64_RELATIVE -
+# RELR-NEXT:   0x30420 R_AARCH64_RELATIVE -
+# RELR-NEXT:   0x30428 R_AARCH64_RELATIVE -
+# RELR-NEXT:   0x30452 R_AARCH64_RELATIVE -
+# RELR-NEXT: }
+
+# RUN: llvm-readobj -x .test %t2 | FileCheck --check-prefix=HEX %s
+
+# HEX:      Hex dump of section '.test':
+# HEX-NEXT: 0x00030410 01000000 2a000020 02000000 2b000000
+##                     ^^^^^^^^ Addend = 1
+##                              ^^^^ Discr = 42
+##                                    ^^ Key (bits 5..6) = DA
+##                                       ^^^^^^^^ Addend = 2
+##                                                ^^^^ Discr = 43
+##                                                      ^^ Key (bits 5..6) = IA
+# HEX-NEXT: 0x00030420 03000000 2c000080 78563412 2d000020
+##                     ^^^^^^^^ Addend = 3
+##                              ^^^^ Discr = 44
+##                                    ^^ Key (bits 5..6) = IA
+##                                    ^^ Addr diversity (bit 7) = true
+##                                       ^^^^^^^^ Addend = 0x12345678
+##                                                ^^^^ Discr = 45
+##                                                      ^^ Key (bits 5..6) = DA
+# HEX-NEXT: 0x00030430 00000000 2e000020 00000000 2f000020
+##                     ^^^^^^^^ No implicit addend (rela reloc)
+##                              ^^^^ Discr = 46
+##                                    ^^ Key (bits 5..6) = DA
+##                                       ^^^^^^^^ Addend = 0
+##                                                ^^^^ Discr = 47
+##                                                      ^^ Key (bits 5..6) = DA
+# HEX-NEXT: 0x00030440 00000000 30000000 00000000 00310000
+##                     ^^^^^^^^ Addend = 0
+##                              ^^^^ Discr = 48
+##                                    ^^ Key (bits 5..6) = IA
+##                                         ^^^^^^ ^^ No implicit addend (rela reloc)
+##                                                  ^^^^ Discr = 49
+# HEX-NEXT: 0x00030450 20000500 00003200 0020{{\ }}
+##                     ^^ Key (bits 5..6) = DA
+##                         ^^^^ ^^^^ Addend = 5
+##                                  ^^^^ Discr = 48
+##                                         ^^ Key (bits 5..6) = DA
+
+.section .test, "aw"
+.p2align 3
+.quad (__ehdr_start + 1)@AUTH(da,42)
+.quad (__ehdr_start + 2)@AUTH(ia,43)
+.quad (__ehdr_start + 3)@AUTH(ia,44,addr)
```

Comment:
Sorry, I just realized that we should not rely on `__ehdr_start` to test non-preemptible symbols. `__ehdr_start` has a value of zero and can mask some errors.
You may change one `__ehdr_start` to `.test` (STT_SECTION).

Reply:
Changed several `__ehdr_start` to `.test`, thanks. Initially, that was taken from [lld/test/ELF/pack-dyn-relocs.s](https://github.com/llvm/llvm-project/blob/main/lld/test/ELF/pack-dyn-relocs.s). In that test, only `__ehdr_start` is also used. I suppose that we might want to replace some of `__ehdr_start` with `.data` (STT_SECTION) there as well. If so, I'll submit a subsequent PR with such changes - please let me know if it's needed.

It turns out that in `processAux` we do not have a final value of `sym.getVA(addend)`, so with a non-zero non-preemptible symbol (e.g., .test section, as you suggested) and a large 32-bit addend, like 0x7fffffff, we might go into a situation when we initially put the reloc to a packed section, but the final implicit value becomes wider than 32 bits. I've fixed that in e9b53374522fca221414aac3d5a3b225e7fc73d4 - see the commit message, comments in code and tests for details, I've tried to describe everything non-trivial.